src/EVM.hs

```diff
@@ -1456,16 +1477,19 @@ notStatic continue = do
     else continue
 
 -- | Burn gas, failing if insufficient gas is available
-burn :: Word64 -> EVM s () -> EVM s ()
-burn n continue = do
+burn :: Gas -> EVM s () -> EVM s ()
+burn amt continue = do
   available <- use (#state % #gas)
-  if n <= available
-    then do
-      #state % #gas %= (subtract n)
-      #burned %= (+ n)
-      continue
-    else
-      vmError (OutOfGas available n)
+  case (amt, available) of
+    (G (Just n), G (Just a)) ->
+      if n <= a
+        then do
+          #state % #gas %= (subtract (G (Just n)))
+          #burned %= (+ (G (Just n)))
+          continue
+        else
+          vmError (OutOfGas a n)
+    _ -> noop
```

Comment:
```suggestion
    _ -> continue
```

It took me almost a day to figure this one out, but I learned so much  :) It's an infinite loop without it.

If you run:
```
:main -p "access-out-of-bounds-array"
```

Then eventually you'll end up in an `OpSstore`, which calls:

```
                      burn (storage_cost + cold_storage_cost) $ do
                        doStore
                        [...]
```

Where `doStore` is:
```
              let doStore = do
                    next
                    assign (#state % #stack) xs
                    modifying (#env % #contracts % ix self % #storage) (writeStorage x new)
```

i.e. it calls `next`, which advances the PC. Without this advancement of the PC, it will loop forever on `OpSstore`. Since `noop` doesn't call `continue`, which in this case is `doStore` (and some other things), which would call `next`, we never advance the PC, and stay inside `OpSstore` forever.